### DISCUSSION TITLE
GraphQL "where" clause: filter using a "Text Field" I have used in all the Content Items

### DISCUSSION THREAD

**Discussion by MarGraz** [OrchardCMS/OrchardCore#15959] [Q&A]
Hi everyone,

I already read a similar topic [here](https://github.com/OrchardCMS/OrchardCore/discussions/12630), but my question is a bit different. I'm working on several headless Orchard Core 1.8.3 microservices, and **I want to be able to query the different Content Items I created, using GraphQL with a `where` clause on a "UserId" `Text field` that all the Content Items I created have.**

I followed the learning path about [GraphQL](https://github.com/Lombiq/Orchard-Training-Demo-Module/blob/dev/Lombiq.TrainingDemo/GraphQL/Startup.cs), and [here](https://github.com/Lombiq/Orchard-Training-Demo-Module/blob/dev/Lombiq.TrainingDemo/GraphQL/Services/PersonPartWhereInputObjectGraphType.cs) it explains how to implement a `WhereInputObjectGraphType`.

The problem is, **I want to avoid defining _n_ `Models` and _n_ `WhereInputObjectGraphType` for each Content Item** because the WhereInputObjectGraphType requires a model.

1. Is there a way to apply the `where` clause on a "UserId" `Text field`, across all the Content Items without defining _n_ models?

2. Otherwise, do you think that the best way to achieve this is to create my custom `SQL Query` and use it through GraphQL? 

3. Or is better to implement an ApiController and use the ISession? But I would lose the advantage of having GraphQL.

4. Perhaps I should use a `Common` part (*) to have only one model? I think that in this case the problem becomes filtering using the correct `Type`, because there isn't a `ContentItemType` in the `where` clause.

Thank you

> (*) To add a `Common` part, go to the Content Type, **add a part** of type `Common`, and save. Now click on the 'Edit' button of the common part, then select 'Part Definition' (on the green banner), and click 'Add Field'. Choose a `Text field` to be used as a common part, and save. Add the same `Common` part to the other Content Types.

**Comment by MarGraz** [comment]
@Piedone do you have any idea of what could be the best solution? 😊 

Thank you

**Comment by Piedone** [comment]
I don't think there's a built-in way to expose content fields in GraphQL for filtering automatically.

We could perhaps extend [Content Fields Indexing (SQL)](https://docs.orchardcore.net/en/latest/reference/modules/SQLIndexing/) with this, since it allows you to filter on content fields from SQL queries. That can also be used in the Queries module, so another workaround would be to expose SQL Queries in GraphQL, what's not available now, [only for Lucene and Elasticsearch queries](https://docs.orchardcore.net/en/main/reference/modules/Queries/#graphql).

**Reply by MarGraz** [reply]
Thank you for your reply.

To extend the [Content Fields Indexing (SQL)](https://docs.orchardcore.net/en/latest/reference/modules/SQLIndexing/) and use it in the `Queries` module, is it sufficient to create a class as explained in the [Usage](https://docs.orchardcore.net/en/latest/reference/modules/SQLIndexing/#usage) section, or do I need to use the `IndexProvider` and implement a `MyTextFieldIndexProvider`?

Is it necessary to register the class showed in the Usage in the `Startup`, or create it in a particular folder? 
It's not clear for me how OC will recognize this new class.

Thank you

**Reply by MarGraz** [reply]
I think I got it, thanks to this [old discussion](https://github.com/OrchardCMS/OrchardCore/discussions/13339#discussioncomment-5215238) where @hishamco replied 😊

The [Content Fields Indexing (SQL)](https://docs.orchardcore.net/en/latest/reference/modules/SQLIndexing/) must be activated as a `feature`, and it will automatically add some new DB tables, like the one related to the `TextFieldIndex`.

After I activated the `Content Fields Indexing (SQL)`, I tried to perform this query:

```sql
SELECT * FROM TextFieldIndex
WHERE ContentType = 'Distributor' AND ContentField = 'UserId'
```

but I got no result 🤔
I saw that **to get results**, it was necessary to save again all my `Distributor` Content Items that I had already created. After that, the query returned a result 🥳 
Thanks to the `DocumentId` I get in the response, I'm able to perform a more complex query and return the entire `Distributor` object.

@Piedone, is this what you intended with "_extend Content Fields Indexing with this, since it allows you to filter on content fields from SQL queries_"?

I think it's necessary to update the [Usage](https://docs.orchardcore.net/en/latest/reference/modules/SQLIndexing/#usage) section in the reference, because **it doesn't mention that is mandatory to enable the Content Fields Indexing (SQL) feature**.

Thank you

**Reply by Piedone** [reply]
While I was referring to querying for this, unfortunately, this won't get you fully there with GraphQL, since this is only about SQL Queries. And the connection to GraphQL is missing, so that's what I think we should add: the easiest would perhaps be to expose all SQL Queries in GraphQL, exactly as what happens for Lucene and Elasticsearch Queries.

I should've actually linked [here](https://docs.orchardcore.net/en/latest/reference/modules/SQLIndexing/#content-fields-indexing), since that's about the module. There it's quite clear that we're talking about a module (what, as all modules, need to be turned on). Somewhat confusingly, the rest of the page is about indexing in general. Perhaps that needs clarification.

**Reply by MarGraz** [reply]
Thank you for your reply.

> the easiest would perhaps be to expose all SQL Queries in GraphQL, exactly as what happens for Lucene and Elasticsearch Queries.

But is not implemented yet in Orchard Core, if I understood well.

> I should've actually linked [here](https://docs.orchardcore.net/en/latest/reference/modules/SQLIndexing/#content-fields-indexing), since that's about the module.

Don't worry, I was focused on the bottom of the page and I didn't see the module paragraph 😅

In the end I think I will implement `ApiController` for all CRUD actions, probably is the fastest way to achieve my goal.

Thank you

**Reply by Piedone** [reply]
Yes, this needs to be added to OC. I opened an issue about it: https://github.com/OrchardCMS/OrchardCore/issues/16008.

**Reply by MarGraz** [reply]
Thank you so much 😊

**Reply by Piedone** [reply]
No problem.
